lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs

```diff
@@ -0,0 +1,186 @@
+            }
+
+            Logger.Log(Level.Info, "OnDriverStart: starting timer to monitor job status. _timer interval: {0}, cancellation detector: {1}", _timerIntervalSec, _cancellationDetector);
+
+            // start timer to monitor cancellation signal
+            _timer = new Timer(_timerIntervalSec * 1000);
+            _timer.Elapsed += OnTimer;
+            _timer.AutoReset = true;
+            _timer.Start();
+        }
+
+        public IDisposable Subscribe(IObserver<IJobCancelled> observer)
+        {
+            if (observer == null)
+            {
+                throw new ArgumentNullException("observer");
+            }
+
+            lock (_observers)
+            {
+                if (!_observers.Contains(observer))
+                {
+                    _observers.Add(observer);
+                }
+            }
+
+            return new AnonymousDisposable(() => Unsubscribe(observer));
+        }
+
+        public void Dispose()
+        {
+            if (_timer != null)
+            {
+                lock (_disposeLock)
+                {
+                    if (_timer != null)
+                    {
+                        _timer.Stop();
+                        _timer.Dispose();
+                        _timer = null;
+                    }
+                }
+            }
+        }
+
+        public void OnError(Exception error)
+        {
+        }
+
+        public void OnCompleted()
+        {
+        }
+
+        private void Unsubscribe(IObserver<IJobCancelled> observer)
+        {
+            lock (_observers)
+            {
+                _observers.Remove(observer);
+            }
+        }
+
+        private void OnTimer(object source, ElapsedEventArgs e)
+        {
+            if (!_observers.Any())
+            {
+                Logger.Log(Level.Warning, "There are no observers for cancellation event: skipping cancellation detection");
+                return;
+            }
+
+            string cancellationMessage = null;
+            if (IsJobCancelled(out cancellationMessage))
+            {
+                Logger.Log(Level.Info, "Detected Job cancellation ({0}): sending JobCancelled event to observers: {1}", cancellationMessage, _observers);
+                var cancelEvent = new JobCancelled(DateTime.Now, cancellationMessage);
+
+                lock (_observers)
+                {
+                    _observers.ForEach(o => o.OnNext(cancelEvent));
+                }
+            }
+        }
+
+        private bool IsJobCancelled(out string cancellationMessage)
+        {
+            var isCancelled = false;
+            cancellationMessage = null;
+
+            try
+            {
+                isCancelled = _cancellationDetector != null && _cancellationDetector.IsJobCancelled(out cancellationMessage);
+            }
+            catch (Exception ex)
+            {
+                Logger.Log(Level.Error, "IsCancelled check failed. Exception:{0}", ex);
```

Comment:
the exception will not do anything as it's on timer thread. 
We could send cancellation signal to the driver and abort the job, but  this might impact lots of users due to a dev error. so I am not sure we should take any drastic steps in this case.